### DISCUSSION TITLE
`JVP/VJP` materializes the full array

### DISCUSSION THREAD

**Discussion by neel04** [jax-ml/jax#24825] [Q&A]
I'm trying to debug excessive memory use when training LMs. It seems that a 65M model with a batch size `> 32` seems to OOM on a 16GB card. The situation is worse for bigger batches and models.

<details><summary>Profiling Graphs:</summary>
<p>

![image](https://github.com/user-attachments/assets/6af80227-130e-4f99-a8b9-e224c1852875)
![image](https://github.com/user-attachments/assets/56d676b9-f1cb-452b-b439-534b46fc6587)
![image](https://github.com/user-attachments/assets/f1514d7d-ea95-46ae-92ce-9e46bc187f40)

</p>
</details> 

One `Array` I'm noticing seems to be this one, for the `jvp` which apears to be of shape `(batch_size * seqlen, vocab_size) = (32 * 512, 50304)`. This is the biggest allocation on the heap, according to the profiler.

Surely this array shouldn't materialize, as for larger models it'd be severely memory inefficient? Is there something I'm missing here or is this expected?

**Comment by jakevdp** [comment]
It's certainly possible to write code for which a JVP would materialize an array of that size. Without seeing what code you're running, it's hard to say any more than that. Can you include a minimal reproduction of the issue?

**Reply by neel04** [reply]
Its just a standard LM training recipie. The code is a bit of a mess atm, but here's the [full context](https://github.com/neel04/ReAct_Jax/blob/72bed9936877b26459449ff4b8af42c5b9f1829f/ReAct/utils/trainer.py#L90) if that helps. FWIW dlwh at [Levanter](https://github.com/stanford-crfm/levanter) faced the same issue.

Here's a MRE I expanded with a bit of help from claude. This isn't exactly what I obtain in the profiler, but appears close enough: [colab](https://colab.research.google.com/drive/15xgHdqG4Wolr6ZOTD2QsynkqXtDAlfYW?usp=sharing) (Check TensorBoard under `memory_viewer` -> `jit_get_loss_and_grads`)
